### PR TITLE
Add {{group-by}} helper

### PR DESCRIPTION
**Pull request by martndemus**

### CODE REVIEW

**Comment by martndemus** [comment]
Should we drop support for 1.13?

File: README.md [hunk 1]
```diff
@@ -26,6 +26,23 @@ Pipes the return values of actions in a sequence of actions.
...
 ```
 
+### Group By
+
+Returns an object where the keys are the unique values of the given property,
+the values are an array with all items of the array that have the same value of
+that property.
+
+```hbs
+{{#each-in (group-by artists 'category') as |category artists|}}
```
Comment (poteto) [inline README.md]: Double quotes please! :)

File: addon/helpers/group-by.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+export default Helper.extend({
+  compute([array, byPath]) {
+    set(this, 'array', array);
+    set(this, 'byPath', byPath);
+
+    return get(this, 'content');
+  },
+
+  array:   null,
```
Comment (poteto) [inline addon/helpers/group-by.js]: Small nitpick but prefer not to align values like this

File: addon/helpers/group-by.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  byPath:  null,
+  content: null,
+
+  byPathDidChange: observer('byPath', function() {
+    let byPath = get(this, 'byPath');
+
+    if (byPath) {
+      defineProperty(this, 'content', computed(`array.@each.${byPath}`, groupFunction));
+    } else {
+      defineProperty(this, 'content', null);
```
Comment (poteto) [inline addon/helpers/group-by.js]: Return an empty object? What happens when you use `each-in` on `null`?

File: tests/integration/helpers/group-by-test.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  ]));
+
+  this.render(hbs`
+    {{~#each-in (group-by array 'category') as |category entries|~}}
+      {{~category~}}
+      {{~#each entries as |entry|~}}{{~entry.name~}}{{~/each~}}
+    {{~/each-in~}}
+  `);
+
+  assert.equal(this.$().text().trim(), 'aabbcd');
```
Comment (poteto) [inline tests/integration/helpers/group-by-test.js]: Assertion messages please! 🤓

**Comment by poteto** [comment]
Hmm, I was looking for an `each-in` polyfill but couldn't find one. Maybe we should make it and use it as a dependency here?

**Comment by martndemus** [comment]
We could polyfill `each-in` with a component that does `{{yield key value}}` for each kv pair.

File: app/helpers/group-by.js [hunk 5]
```diff
@@ -0,0 +1 @@
+export { default, groupBy } from 'ember-functional-helpers/helpers/group-by';
```
Comment (poteto) [inline app/helpers/group-by.js]: Can we also export in https://github.com/DockYard/ember-functional-helpers/blob/remove-compose/addon/index.js please

**Comment by martndemus** [comment]
@homu r+

**Comment by martndemus** [comment]
@homu r+

**Comment by poteto** [comment]
Rebase please

**Comment by martndemus** [comment]
@homu r+

File: addon/helpers/group-by.js (hunk 3 above)
Comment (henrymazza) [inline addon/helpers/group-by.js]: @poteto  Got bitten by that. Helper just fails and something swallows all the errors rendering a blank page. Perhaps best behavior should be discard null groupNames unless a `defaultGroupName` provided, like:

```handlebars
{{each-in (group-by "category" "uncategorized" authors)}}
```

**Pull request merged**